lib/platform-stats.js

```diff
@@ -99,30 +101,136 @@ export const updateDailyStationStats = async (
 
 /**
  * @param {import('pg').Client} pgClient
- * @param {Set<string>} participants
+ * @param {number[]} participantIds
  */
-export const updateDailyParticipants = async (pgClient, participants) => {
-  debug('Updating daily participants, count=%s', participants.size)
-  const ids = await mapParticipantsToIds(pgClient, participants)
+export const updateDailyParticipants = async (pgClient, participantIds) => {
+  debug('Updating daily participants, count=%s', participantIds.length)
   await pgClient.query(`
     INSERT INTO daily_participants (day, participant_id)
     SELECT now() as day, UNNEST($1::INT[]) AS participant_id
     ON CONFLICT DO NOTHING
   `, [
-    ids
+    participantIds
+  ])
+}
+
+/**
+ * @param {import('pg').Client} pgClient
+ * @param {import('./preprocess.js').Measurement[]} honestMeasurements
+ * @param {import('./preprocess.js').Measurement[]} allMeasurements
+ * @param {Map<string, number>} participantMap
+ * @param {object} options
+ * @param {Date} [options.day]
+ */
+export const updateStationAndParticipantDetails = async (
+  pgClient,
+  honestMeasurements,
+  allMeasurements,
+  participantMap,
+  { day = new Date() } = {}
+) => {
+  const statsPerStationParticipant = new Map()
+  const participantSubnetsSet = new Set()
```

Comment:
Do you mean just renaming, or changing the logic?

If renaming, I think it won't be an accurate name, since the set key is the pair of both together

Reply:
Unless we're going on the assumption that two subnets would not share the same participant, which I guess we do for reward anyway right?

Reply:
I was suggesting to rename, but you're right that it would be wrong, because of the compound key

Reply:
Ok, I think this also applies to all/most of the other rename suggestions below, right? Each of them have compound keys. What do you think?

Reply:
I see! I didn't realize at first that this is a participant x subnet set, and not a set of participant subnets. Tricky to name this! What about a different data structure: Instead of using a compound key for storing each seen participant x subnet combination, we can use a nested data structure, which is clearer:

```js
const subnets = []
subnets[id].push(participant)
```

This way we also don't need to parse the compound key later again.

Reply:
I am not sure I understand. Do you mean:

```
const subnets = []
subnets[id].push(subnet)
```

instead? If so, would this be an array of arrays? And the index of the outermost is the participant_id? But we would not be getting all the participants in each round, so we'd have gaps (if that's even possible in JS, not sure). And how would we use this data structure?

Reply:
Hm no I meant it the way I wrote it. We need to keep track of stats per subnet per participant. So for each subnet, you would keep track of the participants that are in that subnet. Thinking about it more, an object of objects could work better for later accessing the data:

```js
const subnets = {}
subnets[subnetId][participantId] = { total: 0, accepted: 0 }
```

Reply:
or am I misunderstanding this part?

Reply:
ping @PatrickNercessian

Reply:
Oops, sorry, I never noticed this response. So we have the two data structures for separate reasons:

```
  const statsPerStationParticipant = new Map()
  const participantSubnetsSet = new Set()
```

1. `statsPerStationParticipant` uses a compound key of `stationId` and `participantId`, and tracks the stats for that pair. It's used to insert into `recent_station_details` (because those columns, along with `day`, are the compound primary key of that table).
2. `participantSubnetsSet` uses a compound key of `participantId` and `subnet`. It's used to insert into `recent_participant_subnets` (because those columns, along with `day`, are the compound primary key of that table)

However, we could potentially use a nested structure for `statsPerStationParticipant`, similar to what you described (but the nested keys being `stationId` and `participantId` instead). But we still need the functionality allowed by `participantSubnetsSet`, and I don't see a good way of matching that structure (aside from another nested object structure, where the value is something arbitrary that we never use).

Also, the reason the code has evolved to these compound string keys instead is that we already do this for the existing `daily_stations` code, which uses a triple compound key. Would we want to also turn this into a 3x nested structure? I guess we are eventually planning to remove this `daily_stations` table and code, so maybe we can ignore it for now?

Reply:
@bajtos and me discussed this and came to the following conclusion:

1. It is cleaner to use traditional nested data structures. There's a high maintenance cost when introducing a novel data structure for a single file.
2. For performance, let's use `Map` instead of `{}`

Example of how the nested data structure works:

```js
const subnets = new Map()

for (const m of measurements) {
  if (!subnets.has(m.inet_group)) {
    subnets.set(m.inet_group, new Set())
  }
  subnets.get(m.inet_group).add(participantId)
}

await pgClient.query(`
  INSERT INTO recent_participant_subnets (day, participant_id, subnet)
  SELECT $1::DATE, unnest($2::int[]), unnest($3::text[])
  ON CONFLICT (day, participant_id, subnet) DO NOTHING
, [
  day,
  [...subnets.entries().flatMap(([subnet, participantIds]) => participantIds.map(() => subnet))],
  [...subnets.entries().flatMap(([_, participantIds]) => participantIds)]
])
```

Reply:
Addressed in https://github.com/filecoin-station/spark-evaluate/pull/311/commits/cee07ab2dc0d95c234ee32d2bd6e1fd80072d72a and https://github.com/filecoin-station/spark-evaluate/pull/311/commits/3bf00da3d431da64b0a5d84e00bdeda05ac04c7d